### DISCUSSION TITLE
Converting between union and Maybe

### DISCUSSION THREAD

**Discussion by divanburger** [odin-lang/Odin#2868] [Ideas/Requests]
When converting from an union to a Maybe I have to write code like this:
```
Session :: struct {
	key: string,
	user_id: Maybe(int)
}

ResultValue :: union {
	int,
	string,
	bool
}

row: map[string]ResultValue
session: Session

user_id, ok := row["user_id"].(int)
session.user_id = ok ? user_id : nil
```

As it is not possible to convert directly from an union to a Maybe. 

I though I could maybe write something like: 
```
session.user_id = row["user_id"].(int) or_else nil
```
That unfortunately gives the error `Cannot convert untyped value 'nil' to 'int' from 'untyped nil'` as it is trying to convert the `nil` directly to the type of the value first without realising that the Maybe supports both. The ternary operator works correctly though.

My proposal is either making the `or_else` not try to convert to the type of the value and instead check the type that is supported by the context (the left-hand side of the assignment in this case), or maybe add syntax to convert from an union with a type assert to a `Maybe`. Maybe something like `session.user_id = row["user_id"].?(int)`. I don't really like this syntax but maybe there is something similar that could work? `session.user_id = row["user_id"].(Maybe(int))` wouldn't make sense as the union could have a `Maybe(int)` in it.

Thoughts?

**Comment by JesseRMeyer** [comment]
We discussed this in the discord and it looks like `or_else` should support this.  Thanks for the report.
